### DISCUSSION TITLE
1.0.1 - Nuget Release

### DISCUSSION THREAD

**Discussion by MattB-msft** [microsoft/PowerPlatform-DataverseServiceClient#284] [General]
### 1.0 Release of Dataverse Service Client.
**Thanks all for your help in getting us to this point... Your feedback over the development and refinement time has been key!** 

Main Highlights: 
Fixed a "Duplicate Key Error" issue created when manually adding some request parameters to an organization request, while also setting the appropriate property in the client 
Fix formatting issue with RequestId and Session Id.
Fix exponential back off issue for WebAPI based calls when retrying based on a throttled connection.
Added implementation for Connection and Configuration constructor. 
    - Allows for a instance of the service client to be created and configured without initialization the client.  The client can then be initialized later using the .Connect() method.
    - This constructor accepts 3 parameters
        - connectionOptions (ConnectionOptions) = Options necessary to create the connection.
        - deferConnection (bool) defaulted to false = When set to true, will configure but not create the connection until .Connect() is called. 
        - serviceClientConfiguration (ConfigurationOptions) defaulted to null = When populated, pre-configures the connection configuration. 
    - If deferred connection is set to true, .Connect() must be called before the client is useable. 
Added Connect() method for use with the deferred behavior of the Connection and Configuration Constructor.
Activated support for adding additional headers to request when using the Configuration Constructor. 
    Note: Dataverse will only allow specific headers to be used. Use of this capability requires foreknowledge of this capability and agreement with Power Platform Dataverse team.

**Full Changelog**: https://github.com/microsoft/PowerPlatform-DataverseServiceClient/compare/0.6.6...1.0.1

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/PowerPlatform-DataverseServiceClient/releases/tag/1.0.1'>1.0.1 - Nuget Release</a>.</em>

**Comment by fowl2** [comment]
Looks like some great improvements! Seems like a lot of my concerns from https://github.com/microsoft/PowerPlatform-DataverseServiceClient/discussions/269 has been addressed 🙏.

Random collection of thoughts:
1. Addition of `DataverseServiceClientBuilder` - is there any functionality beyond calling the constructor? Perhaps this is for future extension? 
2. `ConnectionOptions`
   1. `RequireNewInstance` is equivalent to `useUniqueInstance`? I guess now that this has shipped, it can't be aligned :/
   1. `SkipDiscovery` (seems to default to true?) will suppress WhoAmI requests? Or is "discovery legs" something different?
   2. It'd be great if all the default values could be documented.
1. Looks like `UseWebApi` has graduated from its' previous "do not use: debugging only" status? What is the use case?


PS. Would've been great to have a 1.0-pre or something to gather feedback on the new public API surfaces before :)
PPS. having some trouble with the new ctor, raised https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/288

**Reply by MattB-msft** [reply]
DataverseServiceClientBuilder  is a work in progress, which is why its not exposed right now. 
Regarding RequireNewInstance, Yes we had to maintain both for backward compatibility to CrmServiceClient. 
When you pass a URL for the instance, the skipdiscovery behavior is true.  when you don't, its false.   I will dub check with the team on this. 
Default values are documented :)  just in code and not have made it to the external docs yet ( our doc's team has the work done, we just need to get it merged into the intelisense drops. ) 

UseWebAPI is likely going to get defaulted to false given the longer-term direction we are moving with the underlying interface to powerplatform.   We are keeping it "true" by default now as we work though some behavior issues and perf issues that benefit the platform overall. 

0.6.6 was the pre 1.0  ( it was called out in the docs. ) 

Thanks for your feedback
